### PR TITLE
Fix checking whether a room can be published on creation.

### PR DESCRIPTION
**Pull request by clokep**
#7260 attempted to fix this, but the check was not correct, I'm actually surprised this hasn't caused an errors to appear anywhere, they would be of the form:

`TypeError: 'RoomAlias' object is not iterable`

### CODE REVIEW

**Comment by clokep** [comment]
I spun this out of a type hints PR since it was a real bug fix...

**Comment by reivilibre** [comment]
Does it smell of missing a test if we didn't spot this? :S

File: synapse/config/room_directory.py [hunk 1]
```diff
@@ -20,7 +24,7 @@
 class RoomDirectoryConfig(Config):
     section = "roomdirectory"
 
-    def read_config(self, config, **kwargs):
```
Comment (reivilibre) [inline synapse/config/room_directory.py:23]: `config`? `kwargs` maybe?

File: synapse/config/room_directory.py [hunk 2]
```diff
@@ -47,7 +51,7 @@ def read_config(self, config, **kwargs):
                 _RoomDirectoryRule("room_list_publication_rules", {"action": "allow"})
             ]
 
-    def generate_config_section(self, config_dir_path, server_name, **kwargs):
+    def generate_config_section(self, config_dir_path, server_name, **kwargs) -> str:
```
Comment (reivilibre) [inline synapse/config/room_directory.py:54]:
```suggestion
    def generate_config_section(self, config_dir_path: str, server_name: str, **kwargs) -> str:
```
not sure about the `kwargs` but is that right?

**Review by reivilibre — APPROVED** [review]
Seems good though, thank you :)

File: synapse/config/room_directory.py (hunk 1 above)
Comment (clokep) [reply synapse/config/room_directory.py:23]: I have a separate branch which does all of the `read_config` and `generate_config_section` methods at once. Hope that's OK!

**Comment by clokep** [comment]
> Does it smell of missing a test if we didn't spot this? :S

Yes, there was a test for this but the test had a bug in it (it wasn't publishing the room alias when it was supposed to).

I've pushed some additional tests and cleaned up the old ones a bit.

File: synapse/config/room_directory.py (hunk 1 above)
Comment (reivilibre) [reply synapse/config/room_directory.py:23]: aye that's reet :)

File: tests/handlers/test_directory.py [hunk 3]
```diff
@@ -394,30 +393,24 @@ class TestCreateAliasACL(unittest.HomeserverTestCase):
...
     servlets = [directory.register_servlets, room.register_servlets]
 
-    def prepare(self, reactor, clock, hs):
-        # We cheekily override the config to add custom alias creation rules
-        config = {}
+    def default_config(self):
+        config = super().default_config()
+
+        # We cheekily override the config to add custom alias creation and
+        # room publication rules.
```
Comment (reivilibre) [inline tests/handlers/test_directory.py:400]: but you removed the room publication rules? Guess you changed this comment before invalidating it? :)

File: tests/handlers/test_directory.py [hunk 4]
```diff
@@ -505,10 +522,24 @@ def test_allowed_with_publication_permission(self):
...
-            is_public=False,
+            is_public=True,
             expect_code=200,
         )
 
+    def test_denied_publication_with_invalid_alias(self):
+        """
+        Try to create a room, register an alias for it, and publish it,
+        as a user WITH permission to publish rooms.
+        (This is used as both a standalone test & as a helper function.)
```
Comment (reivilibre) [inline tests/handlers/test_directory.py:533]: Is it?

**Review by reivilibre — APPROVED** [review]
Looks good to me other than these two comments :)

File: tests/handlers/test_directory.py (hunk 3 above)
Comment (clokep) [reply tests/handlers/test_directory.py:400]: Ah, yes. Good catch!

File: tests/handlers/test_directory.py (hunk 4 above)
Comment (clokep) [reply tests/handlers/test_directory.py:533]: Arg, I copied and pasted this, removed it, deleted it all, then redid it. :(

**Pull request merged**